core/index/segment_reader_impl.cpp

```diff
@@ -0,0 +1,346 @@
+
+    return doc_.value;
+  }
+
+  doc_id_t value() const noexcept override { return doc_.value; }
+
+  attribute* get_mutable(irs::type_info::type_id type) noexcept override {
+    return irs::type<document>::id() == type ? &doc_ : nullptr;
+  }
+
+ private:
+  document doc_;
+  doc_id_t max_doc_;  // largest valid doc_id
+};
+
+class MaskDocIterator final : public doc_iterator {
+ public:
+  MaskDocIterator(doc_iterator::ptr&& it, const document_mask& mask) noexcept
+    : mask_{mask}, it_{std::move(it)} {}
+
+  bool next() override {
+    while (it_->next()) {
+      if (!mask_.contains(value())) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  doc_id_t seek(doc_id_t target) override {
+    const auto doc = it_->seek(target);
+
+    if (!mask_.contains(doc)) {
+      return doc;
+    }
+
+    next();
+
+    return value();
+  }
+
+  doc_id_t value() const override { return it_->value(); }
+
+  attribute* get_mutable(irs::type_info::type_id type) noexcept override {
+    return it_->get_mutable(type);
+  }
+
+ private:
+  const document_mask& mask_;  // excluded document ids
+  doc_iterator::ptr it_;
+};
+
+class MaskedDocIterator final : public doc_iterator, private util::noncopyable {
+ public:
+  MaskedDocIterator(doc_id_t begin, doc_id_t end,
+                    const document_mask& docs_mask) noexcept
+    : docs_mask_{docs_mask}, end_{end}, next_{begin} {}
+
+  bool next() override {
+    while (next_ < end_) {
+      current_.value = next_++;
+
+      if (!docs_mask_.contains(current_.value)) {
+        return true;
+      }
+    }
+
+    current_.value = doc_limits::eof();
+
+    return false;
+  }
+
+  doc_id_t seek(doc_id_t target) override {
+    next_ = target;
+    next();
+
+    return value();
+  }
+
+  attribute* get_mutable(irs::type_info::type_id type) noexcept override {
+    return irs::type<document>::id() == type ? &current_ : nullptr;
+  }
+
+  doc_id_t value() const override { return current_.value; }
+
+ private:
+  const document_mask& docs_mask_;
+  document current_;
+  const doc_id_t end_;  // past last valid doc_id
+  doc_id_t next_;
+};
+
+std::vector<index_file_refs::ref_t> GetRefs(const directory& dir,
+                                            const SegmentMeta& meta) {
+  std::vector<index_file_refs::ref_t> file_refs;
+  file_refs.reserve(meta.files.size());
+
+  auto& refs = dir.attributes().refs();
+  for (auto& file : meta.files) {
+    file_refs.emplace_back(refs.add(file));
+  }
+
+  return file_refs;
+}
+
+}  // namespace
+
+/*static*/ std::shared_ptr<const SegmentReaderImpl> SegmentReaderImpl::Open(
+  const directory& dir, const SegmentMeta& meta,
+  const IndexReaderOptions& opts) {
+  auto& codec = *meta.codec;
+
+  auto reader = std::make_shared<SegmentReaderImpl>(dir, meta, opts);
+
+  // read document mask
+  if (opts.doc_mask) {
+    index_utils::ReadDocumentMask(reader->docs_mask_, dir, meta);
+  } else {
+    reader->info_.live_docs_count = reader->info_.docs_count;
```

Comment:
I don't see who set reader->info_.live_docs_count in first branch

Reply:
It's not needed as incoming meta already has correct numbers.

Reply:
I don't understand info_ is not meta¿